spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/Spark3ZOrderStrategy.java

```diff
@@ -0,0 +1,302 @@
+      SparkSession cloneSession = spark().cloneSession();
+      cloneSession.conf().set(SQLConf.ADAPTIVE_EXECUTION_ENABLED().key(), false);
+
+      // Reset Shuffle Partitions for our sort
+      long numOutputFiles = numOutputFiles((long) (inputFileSize(filesToRewrite) * sizeEstimateMultiple()));
+      cloneSession.conf().set(SQLConf.SHUFFLE_PARTITIONS().key(), Math.max(1, numOutputFiles));
+
+      Dataset<Row> scanDF = cloneSession.read().format("iceberg")
+          .option(SparkReadOptions.FILE_SCAN_TASK_SET_ID, groupID)
+          .load(table().name());
+
+      Column[] originalColumns = Arrays.stream(scanDF.schema().names())
+          .map(n -> functions.col(n))
+          .toArray(Column[]::new);
+
+      List<StructField> zOrderColumns = zOrderColNames.stream()
+          .map(scanDF.schema()::apply)
+          .collect(Collectors.toList());
+
+      Column zvalueArray = functions.array(zOrderColumns.stream().map(colStruct ->
+          SparkZOrder.sortedLexicographically(functions.col(colStruct.name()), colStruct.dataType())
+      ).toArray(Column[]::new));
+
+      Dataset<Row> zvalueDF = scanDF.withColumn(Z_COLUMN, SparkZOrder.interleaveBytes(zvalueArray));
+
+      SQLConf sqlConf = cloneSession.sessionState().conf();
+      LogicalPlan sortPlan = sortPlan(distribution, ordering, zvalueDF.logicalPlan(), sqlConf);
+      Dataset<Row> sortedDf = new Dataset<>(cloneSession, sortPlan, zvalueDF.encoder());
+      sortedDf
+          .select(originalColumns)
+          .write()
+          .format("iceberg")
+          .option(SparkWriteOptions.REWRITTEN_FILE_SCAN_TASK_SET_ID, groupID)
+          .option(SparkWriteOptions.TARGET_FILE_SIZE_BYTES, writeMaxFileSize())
+          .option(SparkWriteOptions.USE_TABLE_DISTRIBUTION_AND_ORDERING, "false")
+          .mode("append")
+          .save(table().name());
+
+      return rewriteCoordinator.fetchNewDataFiles(table(), groupID);
+    } finally {
+      manager.removeTasks(table(), groupID);
+      rewriteCoordinator.clearRewrite(table(), groupID);
+    }
+  }
+
+  @Override
+  protected org.apache.iceberg.SortOrder sortOrder() {
+    return Z_SORT_ORDER;
+  }
+
+  static class SparkZOrder {
+
+    private static final byte[] TINY_EMPTY = new byte[Byte.BYTES];
+    private static final byte[] SHORT_EMPTY = new byte[Short.BYTES];
+    private static final byte[] INT_EMPTY = new byte[Integer.BYTES];
+    private static final byte[] LONG_EMPTY = new byte[Long.BYTES];
+    private static final byte[] FLOAT_EMPTY = new byte[Float.BYTES];
+    private static final byte[] DOUBLE_EMPTY = new byte[Double.BYTES];
+
+    static byte[] interleaveBits(Seq<byte[]> scalaBinary) {
+      byte[][] columnsBinary = scala.collection.JavaConverters.seqAsJavaList(scalaBinary)
+          .toArray(new byte[scalaBinary.size()][]);
+      return ZOrderByteUtils.interleaveBits(columnsBinary);
+    }
+
+    private static UserDefinedFunction tinyToOrderedBytesUDF() {
```

Comment:
The udfs here are okay, but I think we'd get better performance overall if we were to use `FunctionCatalog` to expose these to Spark. That way, Spark compile in the method calls.